musmart/melody/contour/step_contour.py

```diff
@@ -0,0 +1,217 @@
+
+        normalized = [
+            self._step_contour_length * (duration / total_duration)
+            for duration in durations
+        ]
+        return normalized
+
+    def _expand_to_vector(
+        self,
+        pitches: list[int],
+        normalized_durations: list[float]
+    ) -> list[int]:
+        """Helper function to create a vector of length step_contour_length by repeating
+        each pitch value proportionally to its normalized duration.
+
+        Parameters
+        ----------
+        pitches : list[int]
+            List of pitch values
+        normalized_durations : list[float]
+            List of normalized duration values (should sum to step_contour_length)
+
+        Returns
+        -------
+        list[int]
+            List of length step_contour_length containing repeated pitch values
+
+        Examples
+        --------
+        >>> sc = StepContour([60], [1.0], step_contour_length=4)
+        >>> sc._expand_to_vector([60, 62], [2.0, 2.0])
+        [60, 60, 62, 62]
+        """
+        result = []
+        for pitch, duration in zip(pitches, normalized_durations):
+            repetitions = round(duration)
+            result.extend([pitch] * repetitions)
+
+        if len(result) > self._step_contour_length:
+            result = result[:self._step_contour_length]
+        elif len(result) < self._step_contour_length:
+            result.extend([result[-1]] * (self._step_contour_length - len(result)))
+
+        return result
+
+    def _calculate_contour(
+        self,
+        pitches: list[int],
+        durations: list[float]
+    ) -> list[int]:
+        """Calculate the step contour from input pitches and durations.
+
+        Examples
+        --------
+        >>> sc = StepContour([60, 62], [1.0, 1.0], step_contour_length=4)
+        >>> sc._calculate_contour([60, 62], [1.0, 1.0])
+        [60, 60, 62, 62]
+        """
+        normalized_durations = self._normalize_durations(durations)
+        return self._expand_to_vector(pitches, normalized_durations)
+
+    @property
+    def global_variation(self) -> float:
+        """Calculate the global variation of the step contour.
+
+        Returns
+        -------
+        float
+            Float value representing the global variation of the step contour
+
+        Examples
+        --------
+        >>> sc = StepContour([60, 62, 64], [1.0, 1.0, 1.0])
+        >>> round(sc.global_variation, 2)
+        1.64
+        """
+        return float(np.nanstd(self.contour))
+
+    @property
+    def global_direction(self) -> float:
+        """Calculate the global direction of the step contour.
+
+        Returns
+        -------
+        float
+            Float value representing the global direction of the step contour
+            Returns 0.0 if the contour is flat
+
+        Examples
+        --------
+        >>> sc = StepContour([60, 62, 64], [1.0, 1.0, 1.0])
+        >>> round(sc.global_direction, 3)
+        0.943
```

Comment:
Where did you get 0.943 from? What makes you think this is right answer? Looking at the input, one might have thought the output should be 1.0....

Reply:
I've validated this with the original implementation in R - because step contour is exactly that, stepped, the correlation with a linearly increasing function is made slightly weaker!